node/as/thrift.js

```
@@ -0,0 +1,171 @@
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var assert = require('assert');
+
+module.exports = TChannelAsThrift;
+
+function TChannelAsThrift(opts) {
+    var self = this;
+    assert(opts.spec, 'TChannelAsThrift expected spec');
+    self.spec = opts.spec;
+}
+
+TChannelAsThrift.prototype.register =
+function register(channel, name, opts, handle) {
+    var self = this;
+
+    var argsName = name + '_args';
+    var argsType = self.spec.getType(argsName);
+
+    var returnName = name + '_result';
+    var resultType = self.spec.getType(returnName);
+
+    channel.handler.register(name, handleThriftRequest);
+
+    function handleThriftRequest(req, res, inHeadBuffer, inBodyBuffer) {
+
+        // Process incoming thrift body
+        var inBodyResult = argsType.fromBuffer(inBodyBuffer);
+        if (inBodyResult.error) {
+            return inBodyResult.toCallback(handleThriftResponse);
+        }
+        var inBody = inBodyResult.value;
+
+        // TODO process inHeadBuffer into inHead
+        var inHead = null;
+
+        handle(opts, inHead, inBody, handleThriftResponse);
+
+        function handleThriftResponse(error, response) {
+            if (error) {
+                return res.sendError('UnexpectedError', error.message);
+            }
+
+            // TODO {head,body} or {arg2,arg3}?
+
+            var ok = response.ok;
+            var outBody = response.body;
+
+            if (typeof ok !== 'boolean') {
+                throw new Error('Expected true or false boolean on response object');
```

If the application developer responds with `ok: undefined` then it's simply ambigious whether its true or false.

Better to fail hard on what is a programmer error ( the application developer did not choose whether the call response is ok ) then to have undefined behaviour.